src/query/plan/operator.cpp

```diff
@@ -5222,4 +5223,133 @@ void IndexedJoin::IndexedJoinCursor::Reset() {
+
+    if (!hash_join_initialized_) {
+      // Pull all left_op_ frames
+      while (left_op_cursor_->Pull(frame, context)) {
+        ExpressionEvaluator evaluator(&frame, context.symbol_table, context.evaluation_context, context.db_accessor,
+                                      storage::View::NEW);
+        auto left_value = self_.hash_join_condition_->expression2_->Accept(evaluator);
+        if (left_value.type() != TypedValue::Type::Null) {
+          hashtable_[left_value].emplace_back(frame.elems().begin(), frame.elems().end());
+        }
+      }
+      hash_join_initialized_ = true;
+    }
+
+    // If left_op yielded zero results, there is no cartesian product.
+    if (hashtable_.empty()) {
+      return false;
+    }
+
+    auto restore_frame = [&frame, &context](const auto &symbols, const auto &restore_from) {
+      for (const auto &symbol : symbols) {
+        frame[symbol] = restore_from[symbol.position()];
+        if (context.frame_change_collector && context.frame_change_collector->IsKeyTracked(symbol.name())) {
+          context.frame_change_collector->ResetTrackingValue(symbol.name());
+        }
+      }
+    };
+
+    if (!common_value_found_) {
+      // Pull from the right_op until there’s a mergeable frame
+      while (true) {
+        auto pulled = right_op_cursor_->Pull(frame, context);
+        if (!pulled) return false;
+
+        // Check if the join value from the pulled frame is shared with any left frames
+        ExpressionEvaluator evaluator(&frame, context.symbol_table, context.evaluation_context, context.db_accessor,
+                                      storage::View::NEW);
+        auto right_value = self_.hash_join_condition_->expression1_->Accept(evaluator);
+        if (hashtable_.contains(right_value)) {
+          // If so, finish pulling for now and proceed to joining the pulled frame
+          right_op_frame_.assign(frame.elems().begin(), frame.elems().end());
+          common_value_found_ = true;
+          common_value = right_value;
+          left_op_frame_it_ = hashtable_[common_value].begin();
+          break;
+        }
+      }
+    } else {
+      // Restore the right frame ahead of restoring the left frame
+      restore_frame(self_.right_symbols_, right_op_frame_);
+    }
+
+    restore_frame(self_.left_symbols_, *left_op_frame_it_);
+
+    left_op_frame_it_++;
+    // When all left frames with the common value have been joined, move on to pulling and joining the next right frame
+    if (common_value_found_ && left_op_frame_it_ == hashtable_[common_value].end()) {
+      common_value_found_ = false;
+    }
+
+    return true;
+  }
+
+  void Shutdown() override {
+    left_op_cursor_->Shutdown();
+    right_op_cursor_->Shutdown();
+  }
+
+  void Reset() override {
+    left_op_cursor_->Reset();
+    right_op_cursor_->Reset();
+    hashtable_.clear();
+    right_op_frame_.clear();
+    hash_join_initialized_ = false;
+    common_value_found_ = false;
+  }
```

Comment:
`left_op_frame_it_` is missing here

Reply:
That’s intended – as `left_op_frame_it_` is an iterator over one value in `hashtable_`, it’s only ever used if `left_op_frame_it_ = hashtable_[common_value].begin();` has been run beforehand. Because of that, there’s no harm if I don’t reset it here.
Also, in `Reset`, if the `hashtable_` is empty to begin with, resetting `left_op_frame_it_` makes no sense because it has never been set.

Reply:
I would reset it just for consistency. If someone later changes the code and wants to check if left_op_frame_it_ is still set, but it should be reset here, it would be a nasty bug.